### PR TITLE
Ingress and Route generation refactor.

### PR DESCRIPTION
**Pull request by dev-gaur**

### CODE REVIEW

File: core/src/main/java/io/fabric8/maven/core/config/ResourceConfig.java [hunk 1]
```diff
@@ -92,7 +92,10 @@
     @Parameter
     private List<ServiceAccountConfig> serviceAccounts;
 
-    private List<IngressRule> ingressRules;
```
Comment (dev-gaur) [inline core/src/main/java/io/fabric8/maven/core/config/ResourceConfig.java]: I added this wrongfully in my previous PR here: https://github.com/fabric8io/fabric8-maven-plugin/pull/1631/commits/dc5b213995d55e7470e08d4f2a9cbf7879546c9a#diff-9238e96d57160d3b61190b609607b2d8R95

hence, removing it

**Comment by dev-gaur** [comment]
@ro14nd @rohanKanojia Do we want the IngressEnricher to be included in the default profile for the default yaml manifests generation, like Route for OpenShift?

Or should there be a flag/mojo parameter for eg., `enableIngress` , `expose` = true to enable invokation of RouteEnricher and IngressEnricher ?

In simpler words, do we want Ingress.yaml to be generated by default for Kubernetes runtime mode?
@manusa

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java [hunk 2]
```diff
@@ -195,6 +195,13 @@
     @Parameter(property = "fabric8.gitRemote")
     private String gitRemote;
 
+
+    /**
+     * The domain added to the service ID when creating OpenShift routes
+     */
+    @Parameter(property = "fabric8.domain")
```
Comment (manusa) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: An entry in the documentation for this property would be nice

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/IngressEnricher.java [hunk 3]
```diff
@@ -48,42 +53,63 @@ public IngressEnricher(MavenEnricherContext buildContext) {
         super(buildContext, "fmp-ingress");
     }
 
+    private String routeDomainPostfix;
+
     @Override
     public void create(PlatformMode platformMode, final KubernetesListBuilder listBuilder) {
+        ResourceConfig resourceConfig = getConfiguration().getResource().orElse(null);
```
Comment (manusa) [inline enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/IngressEnricher.java]: Can be more readable if we make use of the `Optional` capabilities:
```
routeDomainPostfix  = getConfiguration().getResource().map(ResourceConfig::getRouteDomain).orElse(null);
```

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/RouteEnricher.java [hunk 4]
```diff
@@ -48,8 +54,16 @@ public RouteEnricher(MavenEnricherContext buildContext) {
         this.generateRoute = getValueFromConfig(GENERATE_ROUTE, true);
     }
 
+    private String routeDomainPostfix;
+
     @Override
     public void create(PlatformMode platformMode, final KubernetesListBuilder listBuilder) {
+        ResourceConfig resourceConfig = getConfiguration().getResource().orElse(null);
```
Comment (manusa) [inline enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/RouteEnricher.java]: See: https://github.com/fabric8io/fabric8-maven-plugin/pull/1730/files#r352459754

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java (hunk 2 above)
Comment (dev-gaur) [reply plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: https://github.com/fabric8io/fabric8-maven-plugin/pull/1730/files#diff-d71102b809959f854348a4454d17ecebL97

I basically shifted this property from ApplyMojo to ResourceMojo. The Usage for the user should not be affected.

Comment (manusa) [reply plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: Ok.
Anyway, for JKube we should consider having documentation for every Mojo property.

Comment (dev-gaur) [reply plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: Yes, for sure!

File: core/src/main/java/io/fabric8/maven/core/config/ResourceConfig.java (hunk 1 above)
Comment (rohanKanojia) [reply core/src/main/java/io/fabric8/maven/core/config/ResourceConfig.java]: Remove import above too

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java (hunk 2 above)
Comment (rohanKanojia) [reply plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: bdw, I don't see any pre-existing documentation related to this property.

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/IngressEnricher.java [hunk 5]
```diff
@@ -23,19 +23,24 @@
 import io.fabric8.kubernetes.api.model.ServiceBuilder;
 import io.fabric8.kubernetes.api.model.ServicePort;
 import io.fabric8.kubernetes.api.model.ServiceSpec;
-import io.fabric8.kubernetes.api.model.extensions.Ingress;
+import io.fabric8.kubernetes.api.model.extensions.HTTPIngressPath;
```
[the remark this answers is not in the document]
Comment (dev-gaur) [reply enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/IngressEnricher.java]: Oh I thought intellij does that automatically. need to check my ide configuration

**Comment by dev-gaur** [comment]
https://github.com/fabric8io/fabric8-maven-plugin/issues/1374

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/IngressEnricher.java [hunk 6]
```diff
@@ -141,8 +177,39 @@ private boolean isExposedService(Service service) {
...
+    /**
+     * Should we try to create an external URL for the given service?
+     * <p/>
+     * By default lets ignore the kubernetes services and any service which does not expose ports 80 and 443
+     *
+     * @return true if we should create an Ingress for this service.
+     */
+    private boolean shouldCreateExternalURLForService(ServiceBuilder service) {
+        String serviceName = service.getMetadata().getName();
+        if ("kubernetes".equals(serviceName) || "kubernetes-ro".equals(serviceName)) {
```
[the remark this answers is not in the document]
Comment (dev-gaur) [reply enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/IngressEnricher.java]: I think these are the prefixes for the potential services generated by kubernetes system itself. I took it from the apply mojo.

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/IngressEnricher.java [hunk 7]
```diff
@@ -141,8 +177,39 @@ private boolean isExposedService(Service service) {
...
+            return false;
+        }
+        ServiceSpec spec = service.getSpec();
+        List<ServicePort> ports = spec.getPorts();
+        log.debug("Service " + serviceName + " has ports: " + ports);
```
[the remark this answers is not in the document]
Comment (dev-gaur) [reply enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/IngressEnricher.java]: I thought its appropriate for log.debug.

**Comment by dev-gaur** [comment]
We need to deduce a logic for finding/detecting kubernetes-controllers inside the cluster, as they are not a namespaced resource. In minikube, the default nginx-ingress-controller deployment is in the kube-system namespace but is it a globally followed arrangement?

**Pull request merged**